database/database.go

```diff
@@ -14,59 +15,105 @@ import (
 	"github.com/bitcoin-sv/pulse/config"
 )
 
-// DBAdapter defines the interface for a database adapter.
-type DBAdapter interface {
-	Connect(cfg *config.DbConfig) (*sqlx.DB, error)
-	DoMigrations(db *sqlx.DB, cfg *config.DbConfig) error
+// DbAdapter defines the interface for a database adapter.
+type DbAdapter interface {
+	Connect(cfg *config.DbConfig) error
+	DoMigrations(cfg *config.DbConfig) error
+	ImportHeaders(inputFile *os.File, log *zerolog.Logger) (int, error)
+	GetDBx() *sqlx.DB
+}
+
+type dbIndex struct {
+	name string
+	sql  string
 }
 
 func Init(cfg *config.AppConfig, log *zerolog.Logger) (*sqlx.DB, error) {
 	dbLog := log.With().Str("subservice", "database").Logger()
 
-	db, err := Connect(cfg.Db)
+	adapter, err := NewDbAdapter(cfg.Db)
 	if err != nil {
 		return nil, err
 	}
 
-	if err := DoMigrations(db, cfg.Db); err != nil {
+	if err = adapter.Connect(cfg.Db); err != nil {
+		return nil, err
+	}
+
+	if err := adapter.DoMigrations(cfg.Db); err != nil {
 		return nil, err
 	}
 
 	if cfg.Db.PreparedDb {
-		if err := ImportHeaders(db, cfg, &dbLog); err != nil {
+		if err := ImportHeaders(adapter, cfg, &dbLog); err != nil {
 			return nil, err
 		}
 	}
 
-	return db, nil
+	return adapter.GetDBx(), nil
 }
 
-// Connect to the database using the specified adapter.
-func Connect(cfg *config.DbConfig) (*sqlx.DB, error) {
-	adapter, err := NewDBAdapter(cfg)
-	if err != nil {
-		return nil, err
+// NewDbAdapter provides the appropriate database adapter based on the config.
+func NewDbAdapter(cfg *config.DbConfig) (DbAdapter, error) {
+	switch cfg.Engine {
+	case config.DBSqlite:
+		return &SQLiteAdapter{}, nil
+	case config.DBPostgreSql:
+		return &PostgreSqlAdapter{}, nil
+	default:
+		return nil, fmt.Errorf("unsupported database engine %s", cfg.Engine)
 	}
-	return adapter.Connect(cfg)
 }
 
-func DoMigrations(db *sqlx.DB, cfg *config.DbConfig) error {
-	adapter, err := NewDBAdapter(cfg)
+func dropIndexes(db *sqlx.DB, indexQuery *string) (func() error, error) {
```

Comment:
The function returns a closure that calls `restoreIndexes` when invoked. This is an interesting design, giving the caller the option to restore the indexes if needed. Just make sure this behavior is clearly documented. Maybe, a short comment would be helpful.

Reply:
User
I agree that the pattern of a function returning another function, particularly for deferred cleanup or undo actions, is indeed a well-known practice in Go. 

I thought about short comment of this function or sth like this:
```golang
type restoreIndexesFunc func() error

func dropIndexes(db *sqlx.DB, indexQuery *string) (restoreIndexesFunc, error) {
```

But it's a minor, so feel free to resolve this conversation if you will.

Reply:
Removed